### ISSUE TITLE
snap: Failure to resolve media type with resources.GetRemote

### ISSUE THREAD

**Issue by arjunrbery** [gohugoio/hugo#11601]
### What version of Hugo are you using (`hugo version`)?

<pre>
$ hugo version
hugo v0.119.0-b84644c008e0dc2c4b67bd69cccf87a41a03937e+extended linux/amd64 BuildDate=2023-09-24T15:20:17Z VendorInfo=snap:0.119.0
</pre>

### Does this issue reproduce with the latest release?
yes

### Steps to reproduce

1. Create a new repo
2. Add `config.yaml` with the following contents
```
# config.yaml
disableKinds:
- sitemap
- taxonomy
- term
outputs:
  home:
  - html
```
3. Add `layouts\index.html` with the following contents
```
{{/* prebuild/layouts/index.html */}}
<ul>
{{ with resources.GetRemote "https://blog.arb.dev/feed/?t=1698195621" | transform.Unmarshal }}
  {{ range .channel.item }}
    {{ $timestamp := time .pubDate }}
    {{ $link := .link | plainify | htmlUnescape }}
    {{ $postID := replace $link "https://blog.arb.dev/" "" }}
    {{ $postID := replace $postID "/" "-" }}
    {{/* 1. */}} {{ $string := printf "<h1>%s</h1>%s" .title .description }} 
    {{/* 2. */}} {{ $filename := printf "post/%s/%s.md" (urlize $timestamp.Year) (urlize $postID) }} 
    {{/* 3. */}} {{ $resource := resources.FromString $filename $string }} 
    {{/* 4. */}} {{ $file := $resource.RelPermalink }}
    <li><a href="{{$filename}}">{{$filename}}</a></li>
  {{ end }}
{{ end }}
</ul>
```
4. Run `hugo server` and you will see a successfully generated page with links to all of files that have been generated
![image](https://github.com/gohugoio/hugo/assets/20059577/fb0bae18-0870-4861-a52a-0735e0c321dd)

5. Run `hugo` and no child files will be generated and the resulting `index.html` page will be blank

Additional info:
If you print the output of `resources.GetRemote "https://proton.me/blog/feed"  | transform.Unmarshal` on build it generates the following error.
```
map[error calling resources.GetRemote:failed to resolve media type for remote resource "https://blog.arb.dev/feed/?t=1698195621"] 
```

Please note, I'm using `https://blog.arb.dev/feed/?t=1698195621` as a stand in for the this example.

**Comment by jmooring** [comment]
I am unable to reproduce the problem as described.

```text
git clone --single-branch -b hugo-github-issue-11601 https://github.com/jmooring/hugo-testing hugo-github-issue-11601
cd hugo-github-issue-11601
rm -rf resources/ public/ && hugo
```

Then look at the contents of the public directory:

```text
tree public
```

Then examine the published home page:

```text
grep -A6 "<ul>" public/index.html 
```

Unless I'm missing something, this seems like a more of a support issue than a bug, so it would be best to continue the conversation on the [forum](https://discourse.gohugo.io/).

**Comment by jmooring** [comment]
I just noticed that you installed Hugo using the snap package, and now I can recreate the problem. Remove the snap package and install using a [prebuilt binary](https://gohugo.io/installation/linux/#prebuilt-binaries).

**Comment by arjunrbery** [comment]
Thanks for the prompt response. The prebuilt binary resolved the issue.

**Comment by jmooring** [comment]
EDIT 2025-08-30T09:04:13-07:00

The original resource URL is no longer functional. Use `https://www.rassoc.com/gregr/weblog/feed/` for testing.

---

This is a bug, a limitation of the snap package, so I'm going to re-open. I don't see a short term fix, but it https://github.com/gohugoio/hugo/issues/11126 would fix the problem. I can also reproduce the problem with a Docker container. In either case, snap or docker, I can retrieve a remote JSON file without error.

The media type returned by the remote server when retrieving `https://blog.arb.dev/feed/?t=1698195621` is:

```text
Content-Type: application/rss+xml; charset=utf-8
```

Go's built-in media types do not include the above. See <https://go.dev/src/mime/type.go>. So, Go's mime package tries to look at the local system media type db (e.g., on Ubuntu `/etc/mime.types`, `/usr/share/mime/globs2`, etc.), but neither the snap (strictly confined)  nor the docker image have access to those files.

With the snap package, maybe we can get a plug to the required files... need to test. Or go down the #11126 path, which I'd rather avoid for now.

---

No problems when running `hugo server` with either the snap or docker container. When running `hugo server` we load an internal list of media list types, but we don't do that when running `hugo`.

---

As expected, testing with `https://daringfireball.net/thetalkshow/rss` works fine with the snap and docker image. Their server returns:

```text
Content-Type: text/xml; charset=utf-8
```

And that media type is one of Go's built-in media types.

---

I can get it working with docker if I install shared-mime-info when building the image, but that approach doesn't work with the snap; neither does creating a system-files plug to access /etc/mime.types. I think we have to wait for https://github.com/gohugoio/hugo/issues/11126.

**Comment by arjunrbery** [comment]
@jmooring I'll also leave that page up for you all to use for testing. Once I see this close, I'll take down that website.

**Comment by jmooring** [comment]
@bep, My memory is a bit fuzzy on this. We do this when running `hugo server`:

https://github.com/gohugoio/hugo/blob/c4a530f104f2b0d1dd82211b77b603a07db7c181/hugolib/site.go#L398-L406

Is there a downside to doing this with `resources.GetRemote`?

**Comment by arjunrbery** [comment]
Hi @jmooring,

I wanted to see if I can pull down that temporary site since the ticket is now unassigned.

Thanks

**Comment by jmooring** [comment]
@arjunrbery Yes, thanks, please proceed.

**Comment by bep** [comment]
>Is there a downside to doing this with resources.GetRemote?

I don't see any downsides.

**Comment by toby3d** [comment]
I reproduce the same behavior in the built binary v0.129.0 installed via `pacman` (`extra/hugo`) on Arch Linux. Similar error when trying to request Mastodon widget via public API, `Content-Type` which is `application/json; charset=utf-8`:
```gohtml
{{- with resources.GetRemote $url (dict "headers" (dict "Accept" "application/json")) -}}
  {{ with .Err }}
    {{ errorf "error during fetch remote resource: %s" . }}
  {{ else }}
    {{ $data = . | transform.Unmarshal }}
    <figure>
      {{- $data.html | safeHTML -}}
    </figure>
  {{ end }}
{{- else -}}
  {{ errorf "cannot get remote resource %q" $url }}
{{- end -}}
```
```bash
ERROR error during fetch remote resource: error calling resources.GetRemote: failed to resolve media type for remote resource "https://mstdn.io/api/oembed?maxheight=0&maxwidth=1024&url=https%3A%2F%2Fmstdn.io%2F%40toby3d%2F110780380038894309"
```

**Comment by jmooring** [comment]
With the snap package I can get this to work doing both of the following:

1. Adding the served media type to the security allowlist:

    ```toml
    [security]
      [security.http]
        mediaTypes = ['^application/rss\+xml; charset=UTF-8$']
    ```

2. Passing the resource's `.Content` to `transform.Unmarshal` instead of passing the resource itself:

    ```text
    {{ with .Content | transform.Unmarshal }}
    ```

But we shouldn't need to do this. See https://github.com/gohugoio/hugo/issues/11601#issuecomment-1781358500 above.

**Comment by solonovamax** [comment]
> I reproduce the same behavior in the built binary 0.129.0-1 installed via `pacman` (`extra/hugo`) on Arch Linux. Similar error when trying to request Mastodon widget via public API, `Content-Type` which is `application/json; charset=utf-8`:

yep, experiencing the exact same issue. when trying to use github & forgejo apis (github returns a content type of `application/json; charset=utf-8` & forgejo returns a content type of `application/json;charset=utf-8`)
tried looking into why a little, but not entirely sure. it seems that go's logic for handling the mimetypes *should* be mostly correct? and it's not an issue with reading the `/etc/mime.types` file, as go includes a default mapping for json already.

because `ParseMediaType` returns the plain media type, a map of params, and an err, and it's always doing the lookups against the `extensions` map using the plain media type. so the `extensions` map should contain a mapping from `application/json` to `.json`, however it seems for some reason it's unable to find it.

I haven't attached a debugger to it or anything, so there could be some subtle bug in their parsing logic.

I'm doing something slightly similar to jmooring, with
```
{{ with .Value }}
    {{ $data = . | transform.Unmarshal (dict "format" "json") }}
{{ end }}
```
